src/main/scala/eu/proteus/solma/sax/SAXDictionaryPredictOperation.scala

```diff
@@ -42,28 +45,31 @@ class SAXDictionaryPredictOperation[T] extends PredictDataStreamOperation[SAXDic
   override def predictDataStream(
     instance: SAXDictionary,
     predictParameters: ParameterMap,
-    input: DataStream[T]): DataStream[SAXPrediction] = {
+    input: DataStream[(T, Int)]): DataStream[SAXPrediction] = {
 
     val windowSize = instance.getNumberWords()
+    val partitionedStream : KeyedStream[(T, Int), Int] = SAX.toKeyedStream[T](input)
 
-    val predictWindowFunction = new AllWindowFunction[T, SAXPrediction, GlobalWindow] {
+    val predictFunction = new WindowFunction[(T, Int), SAXPrediction, Int, GlobalWindow] {
       override def apply(
+        key: Int,
         window: GlobalWindow,
-        input: Iterable[T],
+        input: Iterable[(T, Int)],
         out: Collector[SAXPrediction]): Unit = {
 
         val freq = new JHashMap[String, AtomicLong]
         input.foreach(w => {
-          freq.putIfAbsent(w.toString, new AtomicLong(0))
-          freq.get(w.toString).incrementAndGet()
+          freq.putIfAbsent(w._1, new AtomicLong(0))
```

Comment:
Why AtomicLong? Since the HashMap is only local, there will be no concurrent access, therefore, i do not think that an atomic increment is needed at all. Or, am i missing anything?